File: heap.java
import java.util.ArrayList;

/**
 * heap
 */
public class heap {
    private ArrayList<Integer> heaplist = new ArrayList<>();
    private int size;
    public heap(){
        heaplist.add(0,0);
        size = 0;
    }
    public void switchUp(int index){
        while(((int)index/2) > 0){
            if (heaplist.get(index) > heaplist.get((int)index/2)){
                int tem = heaplist.get((int)index/2);
                heaplist.set((int)index/2, heaplist.get(index));
                heaplist.set(index,tem);
            }
            index = (int)index/2;
        }
    }
    public void switchDown(int index){
        while((index*2)<= this.size){
            int midpos = this.mid(index);
            if (heaplist.get(index) < heaplist.get(midpos)){
                int tem = heaplist.get(index);
                heaplist.set(index, heaplist.get(midpos));
                heaplist.set(midpos,tem);
            }
            index = midpos;
        }
    }
    public int mid(int i){
        if ((int)(i * 2 + 1) > this.size)
            return (i * 2);
        else{
            if (heaplist.get(i*2) > heaplist.get(i*2+1))
                return (i * 2);
            else
                return (i * 2 + 1);
        }
    }
    public void insert(int k){
        if(!heaplist.contains(k)){
            this.heaplist.add(k);
            this.size++;
            this.switchUp(this.size);
        }else{
            System.out.println("This number is already on the heap");
        }
        
    }
    public void delete(){
        int val = this.heaplist.get(1);
        //System.out.println(val);
        this.heaplist.set(1,this.heaplist.get(heaplist.size()-1));
        size--;
        heaplist.remove(heaplist.size()-1);
        this.switchDown(1);
    }
    public String printheap(){
        String a = this.heaplist.toString();
        a = a.substring(4,a.length());
        a = "[" +a;
        return a;
    }
    public int getRoot(){
        return heaplist.get(1);
    }
}

File: heap.py
#contructor of the class
class heap:
    def __init__ (self):
        self.heapList = [0]
        self.size = 0
    #organice the heap from the final
    def putUp(self,index):
        #the mid of the branch
        #find the right place
        while index // 2 >0:
            #replace the fields
            if self.heapList[index] < self.heapList[index // 2]:
                temp = self.heapList[index // 2]
                self.heapList[index // 2] = self.heapList[index]
                self.heapList[index] = temp
            index = index // 2;

    #organice the heap from the top
    def putDown(self,index):
        while(index *2) <= self.size:
            midpos = self.mid(index)
            #replace
            if self.heapList[index] > self.heapList[midpos]:
                tmp = self.heapList[index]
                self.heapList[index] = self.heapList[midpos]
                self.heapList[midpos] = tmp
            index = midpos

    #insert the new value, recive the value
    def insert(self,k):
        #put the element
        self.heapList.append(k)
        #increace the size
        self.size = self.size + 1
        #organize the new element in the position
        self.putUp(self.size)

    # get the father of a leaf
    def mid(self,index):
        if index * 2 + 1 > self.size:
            return index * 2
        else:
            if self.heapList[index*2] < self.heapList[index*2+1]:
                return index * 2
            else:
                return index * 2 + 1

    #delete the minimun value
    def delete(self):
        #value to delete
        val = self.heapList[1]
        # equal to the last
        self.heapList[1] = self.heapList[self.size]
        # reduce the size
        self.size = self.size - 1
        #remove the element
        self.heapList.pop()
        #organize the heap
        self.putDown(1)

    # organize the entire heap from a list
    def buildHeap(self,alist):
        #initial position
        i = len(alist) // 2
        #put the size
        self.size = len(alist)
        #put the list
        self.heapList = [0] + alist[:]
        #oraganize
        while (i > 0):
            self.putDown(i)
            i = i - 1

    def printheap(self):
        a = self.heapList[1:len(self.heapList)]
        return a
